static_frame/core/frame.py

```diff
@@ -1493,42 +1495,66 @@ def from_sql(cls,
             {name}
             {consolidate_blocks}
         '''
-        row_gen = connection.execute(query)
+        sql_iterable = connection.execute(query)
 
-        own_columns = False
         columns = None
-        if columns_depth == 1:
-            columns = cls._COLUMNS_CONSTRUCTOR(b[0] for b in row_gen.description[index_depth:])
-            own_columns = True
-        elif columns_depth > 1:
-            # use IH: get via static attr of columns const
-            constructor = cls._COLUMNS_HIERARCHY_CONSTRUCTOR.from_labels_delimited
-            labels = (b[0] for b in row_gen.description[index_depth:])
-            columns = constructor(labels, delimiter=' ')
+        own_columns = False
+        filter_row = lambda row: row
+
+        if columns_depth >= 1:
+            labels = (col for (col, *_) in sql_iterable.description[index_depth:])
             own_columns = True
 
+            if columns_depth == 1:
+                if columns_select:
+                    indices, labels = map(tuple, zip(*((i, label) for (i, label) in enumerate(labels) if label in columns_select)))
+                    selector = itemgetter(*indices)
+                    filter_row = lambda row: selector(row)
+
+                columns = cls._COLUMNS_CONSTRUCTOR(labels)
+            else: # > 1
+                # use IH: get via static attr of columns const
+                constructor = cls._COLUMNS_HIERARCHY_CONSTRUCTOR.from_labels_delimited
+                columns = constructor(labels, delimiter=' ')
+
+                if columns_select:
+                    # Have to wait until columns is already constructed in order to filter, since
+                    # `from_labels_delimited` is implementing specific logic to convert the SQL columns
+                    # into an IndexHierarchy, thus making filtering on the raw labels effectively impossible
+                    columns_select = set(columns_select)
+                    sel = columns.isin(columns_select)
+                    columns = columns.loc[sel]
+
+                    selector = itemgetter(*np.where(sel)[0])
+                    filter_row = lambda row: selector(row)
```

Comment:
I have updated `IndexHierarhcy.loc_to_iloc` to support Booleans, so you can now just do `itemgetter(*columns.loc_to_iloc(columns_select))`

Reply:
You can also use the result of `loc_to_iloc` to derive the reduced columns with `columns.iloc`, reducing the number of lookups/comparisons to `columns_select`